osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/AbstractPeriodicMeterReadsCommandExecutor.java

```diff
@@ -1,20 +1,217 @@
+import org.opensmartgridplatform.adapter.protocol.dlms.domain.commands.dlmsobjectconfig.DlmsObjectConfigService;
+import org.opensmartgridplatform.adapter.protocol.dlms.domain.commands.dlmsobjectconfig.DlmsObjectType;
+import org.opensmartgridplatform.adapter.protocol.dlms.domain.commands.dlmsobjectconfig.model.DlmsObject;
+import org.opensmartgridplatform.adapter.protocol.dlms.domain.commands.dlmsobjectconfig.model.DlmsProfile;
+import org.opensmartgridplatform.adapter.protocol.dlms.domain.commands.dlmsobjectconfig.model.Medium;
+import org.opensmartgridplatform.adapter.protocol.dlms.domain.commands.dlmsobjectconfig.model.ProfileCaptureTime;
+import org.opensmartgridplatform.adapter.protocol.dlms.domain.commands.utils.AmrProfileStatusCodeHelper;
+import org.opensmartgridplatform.adapter.protocol.dlms.domain.commands.utils.DlmsHelper;
+import org.opensmartgridplatform.adapter.protocol.dlms.domain.entities.DlmsDevice;
+import org.opensmartgridplatform.adapter.protocol.dlms.domain.entities.Protocol;
+import org.opensmartgridplatform.adapter.protocol.dlms.exceptions.BufferedDateTimeValidationException;
+import org.opensmartgridplatform.adapter.protocol.dlms.exceptions.ProtocolAdapterException;
+import org.opensmartgridplatform.dto.valueobjects.smartmetering.*;
+import org.slf4j.Logger;
+
+import java.time.Duration;
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+import java.util.Date;
+import java.util.List;
+import java.util.Set;
 
 public abstract class AbstractPeriodicMeterReadsCommandExecutor<T, R> extends AbstractCommandExecutor<T, R> {
 
-    public AbstractPeriodicMeterReadsCommandExecutor(final Class<? extends PeriodicMeterReadsRequestDataDto> clazz) {
+    private final AmrProfileStatusCodeHelper amrProfileStatusCodeHelper;
+
+    AbstractPeriodicMeterReadsCommandExecutor(final Class<? extends PeriodicMeterReadsRequestDataDto> clazz,
+                                              final AmrProfileStatusCodeHelper amrProfileStatusCodeHelper) {
         super(clazz);
+        this.amrProfileStatusCodeHelper = amrProfileStatusCodeHelper;
+    }
+
+    /**
+     * Calculates/derives the date of the read buffered DataObject.
+     *
+     * @param ctx             context elements for the buffered object conversion
+     * @param previousLogTime the log time of the previous meter read
+     * @param dlmsHelper      dlms helper object
+     * @return the date of the buffered {@link DataObject} or null if it cannot be determined
+     * @throws ProtocolAdapterException
+     * @throws BufferedDateTimeValidationException in case the date is invalid or null
+     */
+    Date readClock(final ConversionContext ctx,
+                   final Date previousLogTime,
+                   final DlmsHelper dlmsHelper)
+            throws ProtocolAdapterException, BufferedDateTimeValidationException {
+
+        final PeriodTypeDto queryPeriodType = ctx.periodicMeterReadsQuery.getPeriodType();
```

Comment:
the readclock method follows two paths, either it gets the date (I did not touch that logic) or it calculates the date based on the previous value (the fix). The above changes suggests adding several new methods and changing it to a streaming API. We could argue whether that is really necessary.

Reply:
The existing code is not clean because the methods are way to large and do multiple things. I am suggesting ways to make the code cleaner. Do with that what you like.